ocaml/typing/typemod.ml

```diff
@@ -2782,34 +2783,60 @@ and type_structure ?(toplevel = None) funct_body anchor env sstr =
         Misc.fatal_error "kind_abbrev not supported!"
   in
 
+  let force_toplevel =
+    (* A couple special cases are needed for the toplevel:
+
+       - Expressions bound by '_' still escape in the toplevel, because they may
+         be printed even though they are not named, and therefore can't be local
+       - Those expressions and also all [Pstr_eval]s must have types of layout
+         value for the same reason (see the special case in
+         [Opttoploop.execute_phrase]).
+    *)
+    Option.is_some toplevel
+  in
+
   let type_str_item
         env shape_map ({pstr_loc = loc; pstr_desc = desc} as item) sig_acc =
     match Jane_syntax.Structure_item.of_ast item with
     | Some jitem -> type_str_item_jst ~loc env shape_map jitem sig_acc
     | None ->
     match desc with
     | Pstr_eval (sexpr, attrs) ->
-        (* We restrict [Tstr_eval] expressions to representable jkinds to
-           support the native toplevel.  See the special handling of [Tstr_eval]
-           near the top of [execute_phrase] in [opttoploop.ml]. *)
         let expr, sort =
+          (* We could consider allowing [any] here when not in the toplevel,
+             though for now the sort is used in the void safety check. *)
           Builtin_attributes.warning_scope attrs
             (fun () -> Typecore.type_representable_expression
                          ~why:Structure_item_expression env sexpr)
         in
+        if force_toplevel then
+          (* See comment on [force_toplevel]. *)
+          begin match Jkind.Sort.default_to_value_and_get sort with
+          | Value -> ()
+          | Void | Float64 | Float32 | Word | Bits32 | Bits64 ->
+            raise (Error (sexpr.pexp_loc, env, Toplevel_unnamed_nonvalue sort))
+          end;
         Tstr_eval (expr, sort, attrs), [], shape_map, env
-    | Pstr_value(rec_flag, sdefs) ->
-        let force_toplevel =
-          (* Values bound by '_' still escape in the toplevel, because
-             they may be printed even though they are not named *)
-          Option.is_some toplevel
-        in
+    | Pstr_value (rec_flag, sdefs) ->
         let (defs, newenv) =
           Typecore.type_binding env rec_flag ~force_toplevel sdefs in
         let defs = match rec_flag with
           | Recursive -> Typecore.annotate_recursive_bindings env defs
           | Nonrecursive -> defs
         in
+        if force_toplevel then
+          (* See comment on [force_toplevel] *)
+          List.iter (fun vb ->
+            match vb.vb_pat.pat_desc with
+            | Tpat_any ->
```

Comment:
The goal is to find an expression whose let-binding binds no variables, right? I don't think this quite does it. What about

```
let #3L = Int64_u.add #1L #2L
```

? It's a bit silly, but I think that might tickle this case.

Also, what if the pattern does bind boxed variables, but the overall expression is still unboxed? For example

```
let #( s1, s2 ) = #( "hello", "world" )
```

? I don't know exactly how utop operates or what the requirement is here, but I'm worried that `Tpat_any` doesn't quite cover it.

Reply:
> The goal is to find an expression whose let-binding binds no variables, right?

No, the goal is to find any expression that will undergo the transformation done by `Opttoploop.execute_phrase` (see above comment - though if it's not clear I can edit!).  This is exactly things that become `Tstr_eval` or `Tstr_value` with a `Tany` pattern.

Testing quickly, your example does not hit this case in utop.  It does hit an (apparently unrelated) fatal error in the native code toplevel, which looks to be related to how `Stdlib_upstream_compatible` is linked - I will make a ticket for that and ask @dkalinichenko-js to take a look.

Reply:
(To be clearer: All other cases, including the unboxed tuple one, should already be rejected by the typechecker if they are problematic.  This bug is specifically about that transformation done by the native code toplevel.)